#4696 Column support for SelectCheckboxMenu

ltlBeBoy opened the pull request:
Fix #4696: I used `SelectOneMenu` as template to add column support for `SelectCheckboxMenu`. A demo is added to the showcase which almost works. But there is a problem: The first column is rendered outside of the checkbox menu. Can you please help me to fix this?

Fix #7406

ltlBeBoy committed:
Added support for columns for SelectCheckboxMenu (#4696)

ltlBeBoy committed:
Merge branch 'PF_4696-SelectCheckboxMenu_column' of https://github.com/ltlBeBoy/primefaces into PF_4696-SelectCheckboxMenu_column

melloware commented on primefaces/src/main/resources/META-INF/primefaces-p.taglib.xml:
```diff
@@ -22889,6 +22889,14 @@
         </attribute>
+        <attribute>
+            <description>
+                <![CDATA[]]>
```
We should add a description here is possible as IntelliJ apparently will ignore fields without a description.

jepsar reviewed (changes requested):
Please add documentation and a what's new entry.

melloware commented:
I will try and take a look and see why its creating the first column outside the menu

melloware commented:
@ltlBeBoy please add the following at the bottom of the renderer.   This tells it not to render its children which is why it was rendering that column outside of the context.

```java
@Override
public void encodeChildren(FacesContext facesContext, UIComponent component) throws IOException {
    //Rendering happens on encodeEnd
}

@Override
public boolean getRendersChildren() {
    return true;
}
```

melloware commented:
Also check your JS DOC errors... `mvn clean package -Pjsdoc` in the primfaces dir.

```
Type SelectCheckboxMenu declares property <labels>, but no such property is used in the code. Remove the doc comment, or use @forcedProp instead of @prop if you really want to include it.
```

ltlBeBoy committed:
Extended support for columns for SelectCheckboxMenu (#4696)

ltlBeBoy committed:
added two SelectCheckboxMenu demos to showcase; improved CountryService in showcase for reproducible country ID; added documentation

ltlBeBoy committed:
whatsnew.md

melloware committed:
Merge branch 'master' into PF_4696-SelectCheckboxMenu_column

ltlBeBoy commented:
OK, a first version is ready:

1. Added support for custom content in overlay panel.
2. Improved filtering according to #9629.
3. Added two demos to the showcase; one with "raw" entities and one with grouped `SelectItem`s.
4. Improved `org.primefaces.showcase.service.CountryService` to set reproducible IDs for `org.primefaces.showcase.domain.Country` (required for `org.primefaces.showcase.convert.CountryConverter`)
5. Extended documentation.
6. Entry for "whatsnew.md".

In my project I also added the following CSS to the theme (accordingly to `SelectOneMenu`). I guess this should be done by PrimeTek too:
```
/*
 * Columns for SelectCheckboxMenu
 */
body {
  .ui-selectcheckboxmenu-panel {
    .ui-selectcheckboxmenu-items {
      &.ui-selectcheckboxmenu-table {
        > thead tr th {
          background: $panelHeaderBg;
          border: $datatableHeaderCellBorder;
          color: $panelHeaderTextColor;
          font-weight: $panelHeaderFontWeight;
        }
      }

      .ui-selectcheckboxmenu-item {
        background: $datatableBodyRowBg;
        color: $datatableBodyRowTextColor;
        border: $datatableBodyRowBorder;

        > td {
          padding: $datatableBodyCellPadding;
        }

        &:first-child {
          border-top-color: transparent;
        }
      }
    }
  }
}
```

A few things that I noticed during the modifications (because I used `SelectOneMenu` as template):

1. `SelectOneMenu` and `SelectCheckboxMenu` now both support dynamic rendering and custom content in overlay panel. But there is one problem: These functions can not be used simultaneously because the custom content is always rendered by JSF! Shouldn't there be an exception or at least a logged warning that it's impossible to activate both at the same time?
2. `SelectOneMenu` fails to render `SelectItemGroup` children when custom content is used. Are you interested that I create another pull request?
3. `SelectOneMenu` fails to render custom content correctly when `omnifaces.GenericEnumConverter` is used. Are you interested that I create another pull request?
4. Is there a use case where a `SelectItem` with `noSelectionOption="true"` makes sense for `SelectCheckboxMenu`?
5. `SelectOneMenu` also supports RTL while `SelectCheckboxMenu` does not. Is that a requirement?
6. `SelectOneMenu` and `SelectCheckboxMenu` both have a variable to specified the height of the overlay panel. But at `SelectOneMenu` it's a `String` which is called `height`. And at `SelectCheckboxMenu` it's an `Integer` which is called `scrollHeight`...
7. `SelectOneMenu` and `SelectCheckboxMenu` both use a variable named `panelId` on the client side. But at `SelectOneMenu` it's the escaped client ID. And at `SelectCheckboxMenu` it's the "raw" client ID...

melloware commented:
Responses...

0. Please open another ticket with the CSS changes you need PrimeTek to make to the theme so I can mark is as "theme" and assign to PrimeTek
1. I don't understand this one and why thy both can't be used at the same time?  Maybe a ticket and a reproducer?
2. Open a ticket please. and then submit your PR against this ticket for review
3. Open a ticket please. and then submit your PR against this ticket for review
4. NO this use case makes no sense for CheckboxMenu I don't think
5. NO not a requirement just leave the TODO in there.  Not all components support RTL fully
6. It should be a `string` not an `integer` so it can take any CSS like "100vh" "100%" "100px" just make sure the code respects if its only a number it add "px" to it. I think you can find examples other places
7. It should probably be consistent?

ltlBeBoy committed:
Code consistency and comments; DRY

ltlBeBoy committed:
Merge branch 'PF_4696-SelectCheckboxMenu_column' of https://github.com/ltlBeBoy/primefaces into PF_4696-SelectCheckboxMenu_column

ltlBeBoy commented:
@melloware 4. - 7. is done. For 1. - 3. I will open a ticket next week.

melloware commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenuRenderer.java:
```diff
@@ -318,19 +333,343 @@ protected void encodeMenuIcon(FacesContext context, SelectCheckboxMenu menu, boo
+            rowStyleClass += checked ? " ui-selectcheckboxmenu-checked" : " ui-selectcheckboxmenu-unchecked";
+            if (selectItem.isNoSelectionOption()) {
+                rowStyleClass += " ui-noselection-option";
+            }
```
if (selectItem.isDisabled()) {
                rowStyleClass += " ui-state-disabled";
            }

jepsar commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenu.java:
```diff
@@ -84,6 +96,19 @@ public String getDefaultEventName() {
     }
 
+    public List<Column> getColumns() {
+        List<Column> columns = new ArrayList<>(5);
```
Do we ever set initial capacity?

melloware replied:
No I think we only do it if we know the number of items specifically.

ltlBeBoy replied:
@jepsar It's copy-paste from `SelectOneMenu`. But I will remove it.

jepsar commented on primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.selectcheckboxmenu.js:
```diff
@@ -121,9 +122,12 @@ PrimeFaces.widget.SelectCheckboxMenu = PrimeFaces.widget.BaseWidget.extend({
-        this.inputs = this.jq.find(':checkbox');
-        this.panelId = this.id + '_panel';
+        //Only include checkboxes with an ID that starts like ID of the menu!
+        this.inputs = this.jq.find(':checkbox').filter('[id^="' + this.id.replace(/:/g,"\\:") + '"]');
```
This is dangerous as the separator might be configured to a different character.

ltlBeBoy replied:
OK. Before it was `this.inputs = this.jq.find(':checkbox');`. But it's not exact enough because there is a toggler checkbox. Maybe I also try by class.

ltlBeBoy replied:
It should be safe to search in the "hidden" `div` that contains the actual checkbox input elements:

`this.inputs = this.jq.filter('.ui-helper-hidden').find(':checkbox');`

What do you think?

jepsar replied:
Better!

ltlBeBoy replied:
@jepsar FYI:
My proposed fix did not work and resulted in failed integration tests. Apparently because for the target element the CSS `display` is set to `none` initially which makes it invisible to the filter.
So I used this instead:
`this.jq.children('div.ui-helper-hidden:not(.ui-input-overlay)').children(':checkbox');`

jepsar commented on primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.selectcheckboxmenu.js:
```diff
@@ -194,31 +197,39 @@ PrimeFaces.widget.SelectCheckboxMenu = PrimeFaces.widget.BaseWidget.extend({
 
         if (this.cfg.scrollHeight) {
-            this.itemContainerWrapper.height(this.cfg.scrollHeight);
+            this.itemContainerWrapper.css('max-height', this.cfg.scrollHeight);
```
Are these height changes relevant to column support? Looks like a different issue to me.

ltlBeBoy replied:
The problem is the code consistency. I used `SelectOneMenuRenderer` as template. There the height property is used to set the max height (when rendering custom content):

```
protected void encodePanel(FacesContext context, SelectOneMenu menu, List<SelectItem> selectItems) throws IOException {
        ...
        String height = null;
        try {
            height = Integer.parseInt(menu.getHeight()) + "px";
        }
        catch (NumberFormatException e) {
            height = menu.getHeight();
        }
        ...
        writer.startElement("div", null);
        writer.writeAttribute("class", SelectOneMenu.ITEMS_WRAPPER_CLASS, null);
        writer.writeAttribute("style", "max-height:" + height, null);
        ...
```

According to the documentation the similar property for `SelectCheckboxMenu` is `scrollHeight` which is also handed over to the client side in `SelectCheckboxMenuRenderer`:

```
protected void encodeScript(FacesContext context, SelectCheckboxMenu menu) throws IOException {
        WidgetBuilder wb = getWidgetBuilder(context);
        wb.init("SelectCheckboxMenu", menu)
                .callback("onShow", "function()", menu.getOnShow())
                .callback("onHide", "function()", menu.getOnHide())
                .callback("onChange", "function()", menu.getOnchange())
                .attr("scrollHeight", menu.getScrollHeight(), Integer.MAX_VALUE) // now attr("scrollHeight", getMaxScrollHeight(menu), null)
```

But the filter function in `forms.selectcheckboxmenu.js` updates the CSS "height" attribute which would result in ignoring the property `scrollHeight`:

```
if(this.itemContainer.height() < this.cfg.initialHeight) {
    this.itemContainerWrapper.css('height', 'auto');
}
else {
    this.itemContainerWrapper.height(this.cfg.initialHeight);
}
```

This is why I choosed using CSS "max-height".

jepsar commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenuRenderer.java:
```diff
@@ -111,8 +127,7 @@ protected void encodeInputs(FacesContext context, SelectCheckboxMenu menu, List<
         int idx = -1;
-        for (int i = 0; i < selectItems.size(); i++) {
-            SelectItem selectItem = selectItems.get(i);
+        for (SelectItem selectItem : selectItems) {
```
https://github.com/primefaces/primefaces/wiki/Contributing-to-PrimeFaces#performance-guidelines

jepsar commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenuRenderer.java:
```diff
@@ -318,19 +333,343 @@ protected void encodeMenuIcon(FacesContext context, SelectCheckboxMenu menu, boo
+        if (menu.getPanelStyleClass() != null) {
+            panelStyleClass += " " + menu.getPanelStyleClass();
+        }
+        // TODO: "RTLAware"!
```
There is still a TODO here

ltlBeBoy replied:
Yes, I asked @melloware if RTL support is a requirement. He said no, but I should keep the "TODO" comment.

melloware replied:
Yep I have no idea what RTL should do in the case of a SelectCheckBoxMenu so I said leave it as a TODO.

ltlBeBoy committed:
Merge branch 'primefaces:master' into PF_4696-SelectCheckboxMenu_column

ltlBeBoy committed:
Fix #9677; Fix #MYFACES-3130; Code consistency

ltlBeBoy committed:
Improved search for checkbox input elements

ltlBeBoy committed:
Fixed Chromium incompatibility when searching input elements; removed faulty copy-paste code; added comments

ltlBeBoy committed:
Fix #9691

ltlBeBoy committed:
Merge branch 'primefaces:master' into PF_4696-SelectCheckboxMenu_column

jepsar commented on docs/13_0_0/components/selectcheckboxmenu.md:
```diff
@@ -90,6 +90,38 @@ function customFilter(itemLabel, filterValue) {
 
+## Custom Content
+SelectCheckboxMenu can display custom content in overlay panel by using column component and the
+var option to refer to each item. Facets for column _header_ may also be used.
```
```suggestion
var option to refer to each item. Facets for column `header` may also be used.
```

jepsar commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenuRenderer.java:
```diff
@@ -318,19 +335,348 @@ protected void encodeMenuIcon(FacesContext context, SelectCheckboxMenu menu, boo
+                });
+
+        //toggler
+        encodeCheckbox(context, null, false, !notChecked, null, "Select All");
```
Is this UI text? Should be loaded from a message bundle. https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/org/primefaces/Messages.properties#L11

melloware replied:
Looks like we have `primefaces.datatable.aria.HEADER_CHECKBOX_ALL = Select All` label https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/org/primefaces/Messages_en.properties

jepsar replied:
Yeah, I already linked it (including line number) 😁

jepsar commented on primefaces/src/main/java/org/primefaces/component/selectcheckboxmenu/SelectCheckboxMenuRenderer.java:
```diff
@@ -318,19 +335,348 @@ protected void encodeMenuIcon(FacesContext context, SelectCheckboxMenu menu, boo
+        writer.startElement("a", null);
+        writer.writeAttribute("class", SelectCheckboxMenu.CLOSER_CLASS, null);
+        writer.writeAttribute("href", "#", null);
+        writer.writeAttribute(HTML.ARIA_LABEL, "Close", null);
```
https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/org/primefaces/Messages.properties#L19

melloware replied:
`MessageFactory.getMessage(InputRenderer.ARIA_FILTER)` for example if you look in the code for `MessageFactory`

melloware replied:
Looks like we have a `primefaces.messages.aria.CLOSE = Close` label: https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/resources/org/primefaces/Messages_en.properties

jepsar commented on primefaces/src/main/resources/META-INF/primefaces-p.taglib.xml:
```diff
@@ -22851,11 +22851,11 @@
             <name>scrollHeight</name>
             <required>false</required>
-            <type>java.lang.Integer</type>
+            <type>java.lang.String</type>
```
We should mention this in the migration guide, or accept Object. @melloware WDYT?

jepsar replied:
@melloware so not change to Object?

melloware replied:
I think he made it consistent with how SelectOne Menu is so we should make it consistent. And since its a change in property type we should mention it in 13.0/migration guide.

jepsar committed:
Update docs/13_0_0/components/selectcheckboxmenu.md

ltlBeBoy committed:
Merge branch 'primefaces:master' into PF_4696-SelectCheckboxMenu_column

ltlBeBoy committed:
localized ARIA labels; extended migration documentation
